pkg/cache/resource_node.go

```diff
@@ -0,0 +1,141 @@
+package cache
+
+import (
+	"maps"
+
+	"sigs.k8s.io/kueue/pkg/resources"
+)
+
+type ResourceNode struct {
+	// Quotas are the ResourceQuotas specified for the current
+	// node.
+	Quotas map[resources.FlavorResource]*ResourceQuota
+	// SubtreeQuota is the sum of the node's quota, as well as
+	// resources available from its children, constrained by
+	// LendingLimits.
+	SubtreeQuota resources.FlavorResourceQuantities
+	// Usage is the quantity which counts against this node's
+	// SubtreeQuota. For ClusterQueues, this is simply its
+	// usage. For Cohorts, this is the sum of child usage past
+	// guaranteedQuotas.
+	Usage resources.FlavorResourceQuantities
+}
+
+func NewResourceNode() ResourceNode {
+	return ResourceNode{
+		Quotas:       make(map[resources.FlavorResource]*ResourceQuota),
+		SubtreeQuota: make(resources.FlavorResourceQuantities),
+		Usage:        make(resources.FlavorResourceQuantities),
+	}
+}
+
+func (r ResourceNode) Clone() ResourceNode {
+	return ResourceNode{
+		Quotas:       maps.Clone(r.Quotas),
+		SubtreeQuota: maps.Clone(r.SubtreeQuota),
+		Usage:        maps.Clone(r.Usage),
+	}
+}
+
+// guaranteedQuota is the capacity which will not be lent the node's
+// Cohort.
+func (r ResourceNode) guaranteedQuota(fr resources.FlavorResource) int64 {
+	if lendingLimit := r.Quotas[fr].lendingLimitSafe(); lendingLimit != nil {
+		return max(0, r.SubtreeQuota[fr]-*lendingLimit)
+	}
+	return 0
+}
+
+type wiredResourceNode[T any] interface {
+	getResourceNode() ResourceNode
+
+	HasParent() bool
+	Parent() T
+}
+
+// available determines how much capacity remains for the current
+// node, taking into account usage and BorrowingLimits.
+//
+// This function may return a negative number in the case of
+// overadmission - e.g. capacity was removed or the node moved to
+// another Cohort.
+//
+// We add an option to ignore the borrowing limit, to support
+// features.MultiplePreemptions=false path. This will be cleaned up in
+// v0.10, when we delete legacy logic.
+func available[T wiredResourceNode[T]](node wiredResourceNode[T], fr resources.FlavorResource, enforceBorrowLimit bool) int64 {
+	r := node.getResourceNode()
+	if !node.HasParent() {
+		return r.SubtreeQuota[fr] - r.Usage[fr]
+	}
+	localAvailable := max(0, r.guaranteedQuota(fr)-r.Usage[fr])
+	parentAvailable := available(node.Parent(), fr, enforceBorrowLimit)
+
+	if borrowingLimit := r.Quotas[fr].borrowingLimitSafe(); enforceBorrowLimit && borrowingLimit != nil {
+		storedInParent := r.SubtreeQuota[fr] - r.guaranteedQuota(fr)
+		usedInParent := max(0, r.Usage[fr]-r.guaranteedQuota(fr))
+		withMaxFromParent := storedInParent - usedInParent + *borrowingLimit
+		parentAvailable = min(withMaxFromParent, parentAvailable)
+	}
+	return localAvailable + parentAvailable
```

Comment:
doesn't this sum double count localAvailable twice?

Reply:
I don't think so. The localAvailable is just the remaining guaranteed quota. Then we need to do some additional math for the effective borrowing limit (or amount of resource that can flow through parent->child edge), given that the child stores quota in its parent